File: DynamoDB-SDK-Examples/node.js/WorkingWithItems/put-item-conditional.js
/* This is an example of a simple PutItem with a ConditionExpression using
the higher level DocumentClient for Amazon DynamoDB. The conditional
expression must be true for this Put call to succeed. */

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand } = require("@aws-sdk/lib-dynamodb");

async function putItems() {
  const client = new DynamoDBClient({ region: "us-west-2" });
  const ddbDocClient = DynamoDBDocumentClient.from(client);
  return await ddbDocClient.send(
      new PutCommand({
        TableName: "RetailDatabase",
        Item: {
          pk: "jim.Robert@somewhere.com", // Partition key
          sk: "metadata",                 // Sort key
          name: "Jim Roberts",
          first_name: "Jim",
          last_name: "Roberts",
          address: {
            road: "456 Nowhere Lane",
            city: "Langely",
            state: "WA",
            pcode: "98260",
            country: "USA",
          },
          username: "jrob",
        },
        // Start conditional expression to make sure the value of the sk attribute begins with the string "meta".
        // If not, then the put will fail with a "ConditionalCheckFailedException"
        ConditionExpression: "begins_with(sk, :val)",
        ExpressionAttributeValues: {":val" : "meta"}
      })
  );
}

putItems()
    .then((data) =>
        console.log("PutItem succeeded:", JSON.stringify(data, null, 2)))
    .catch((error) => console.error(JSON.stringify(error, null, 2)));

File: DynamoDB-SDK-Examples/dotnet/WorkingWithItems/PutItemConditional.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

using Amazon;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;

namespace DotnetSamples.WorkingWithItems
{
    public class PutItemConditional
    {
        private readonly IAmazonDynamoDB amazonDynamoDB;

        public PutItemConditional()
        {
            amazonDynamoDB = new AmazonDynamoDBClient(RegionEndpoint.USWest2);
        }

        public PutItemConditional(IAmazonDynamoDB amazonDynamoDB)
        {
            this.amazonDynamoDB = amazonDynamoDB;
        }

        public async Task ServiceClientExampleAsync()
        {
            try
            {
                var request = new PutItemRequest
                {
                    TableName = "RetailDatabase",
                    Item = new Dictionary<string, AttributeValue>
                    {
                        { "pk", new AttributeValue("jim.bob@somewhere.com") },
                        { "sk", new AttributeValue("metadata") },
                        { "name", new AttributeValue("Jim Bob") },
                        { "first_name", new AttributeValue("Jim") },
                        { "last_name", new AttributeValue("Bob") },
                        {
                            "address",
                            new AttributeValue
                            {
                                M = new Dictionary<string, AttributeValue>
                                {
                                    { "road", new AttributeValue("456 Nowhere Lane") },
                                    { "city", new AttributeValue("Langely") },
                                    { "state", new AttributeValue("WA") },
                                    { "pcode", new AttributeValue("98260") },
                                    { "country", new AttributeValue("USA") },
                                }
                            }
                        },
                        { "username", new AttributeValue("jbob") },
                    },
                    // this condition expression will not allow updates, it will only succeed if the record does not already exist
                    ConditionExpression = "attribute_not_exists(sk)",
                };

                var response = await amazonDynamoDB.PutItemAsync(request);
                Console.WriteLine($"PutItemConditional succeeded.");
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e.Message);
                throw;
            }
        }
    }
}
